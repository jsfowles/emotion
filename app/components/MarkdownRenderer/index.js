import React from 'react';
import ReactMarkdown from 'react-markdown';

import {
  H1,
  H2,
  H3,
  H4,
  H5,
  H6,
  P,
  Ul,
  Li,
  Strong,
  Hr,
  TextWrapper,
} from './styles';

import CodeBlock from '@containers/CodeBlock';

const Headers = ({ level, children }) => {
  const levels = {
    1: <H1>{ children }</H1>,
    2: <H2>{ children }</H2>,
    3: <H3>{ children }</H3>,
    4: <H4>{ children }</H4>,
    5: <H5>{ children }</H5>,
    6: <H6>{ children }</H6>,
  };

  return levels[level] || levels['1'];
};

export default ({ source }) => (
  <TextWrapper>
    <ReactMarkdown
      source={ source }
      renderers={{
        heading: Headers,
        paragraph: ({ children }) => <P>{ children }</P>,
        link: ({ href, children }) => <a href={ href }>{ children }</a>,
        list: ({ ordered, children}) => <Ul>{ children }</Ul>,
        listItem: ({ children }) => <Li>{ children }</Li>,
        thematicBreak: () => <Hr />,
        strong: ({ children }) => <Strong>{ children }</Strong>,
        code: CodeBlock,
      }}
    />
  </TextWrapper>
);

