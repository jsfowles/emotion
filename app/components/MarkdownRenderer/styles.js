import glamorous from 'glamorous';

import TypeScale from '@identity/type';
import { remCalc } from '@lib/helperMethods';
import * as wrappers from '@identity/wrappers';
import * as spacingCalculator from '@identity/spacing';
import { COLORS, MEASURE, FONT_WEIGHT_NORMAL } from '@identity/constants';

export const P = glamorous(TypeScale.P)({
  ...spacingCalculator.spacing({ margin: 'bottom-md' }),
  ':last-child': { marginBottom: 0 },
});

export const H1 = glamorous(TypeScale.H1)({
  ...spacingCalculator.spacing({ margin: 'bottom-xs' }),
  ':last-child': { marginBottom: 0 },
});

export const H2 = glamorous(TypeScale.H2)({
  ...spacingCalculator.spacing({ margin: 'bottom-xs' }),
  ':last-child': { marginBottom: 0 },
});

export const H3 = glamorous(TypeScale.H3)({
  ...spacingCalculator.spacing({ margin: 'bottom-xs' }),
  ':last-child': { marginBottom: 0 },
});

export const H4 = glamorous(TypeScale.H4)({
  ...spacingCalculator.spacing({ margin: 'bottom-xs' }),
  ':last-child': { marginBottom: 0 },
});

export const H5 = glamorous(TypeScale.H5)({
  ...spacingCalculator.spacing({ margin: 'bottom-xs' }),
  ':last-child': { marginBottom: 0 },
});

export const H6 = glamorous(TypeScale.H5)({
  ...spacingCalculator.spacing({ margin: 'bottom-xs' }),
  ':last-child': { marginBottom: 0 },
});

export const Ul = glamorous.ul(
  spacingCalculator.spacing({ margin: 'bottom-md' }),
  {
    paddingLeft: remCalc(20),
    maxWidth: `calc(${MEASURE.MEDIUM} - ${remCalc(20)})`,
  },
);

export const Li = glamorous(TypeScale.Li)({
  listStyleType: 'disc',
});

export const Strong = glamorous.strong({
  fontWeight: FONT_WEIGHT_NORMAL,
});

export const Hr = glamorous.hr(
  spacingCalculator.spacing({ margin: 'bottom-md' }),
  {
    border: 0,
    height: 1,
    background: COLORS.black008,
  },
);

export const TextWrapper = glamorous(wrappers.TextWrapper)({});
