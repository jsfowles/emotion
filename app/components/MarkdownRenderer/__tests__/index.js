import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import MarkdownRenderer from '@components/MarkdownRenderer';

const defaultProps = {
  source: `
# Hello World

Goto [Underbelly](https://underbelly.is)
`,
};

const generateComponent = (additionalProps = {}) => {
  const newProps = {
    ...defaultProps,
    ...additionalProps,
  };
  return <MarkdownRenderer { ...newProps } />;
};

it('renders with props', () => {
  const wrapper = shallow(generateComponent());
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});
