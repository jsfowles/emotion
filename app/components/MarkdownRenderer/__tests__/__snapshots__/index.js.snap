// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with props 1`] = `
<glamorous(glamorous(div))>
  <ReactMarkdown
    escapeHtml={true}
    renderers={
      Object {
        "blockquote": [Function],
        "code": [Function],
        "heading": [Function],
        "inlineCode": [Function],
        "link": [Function],
        "list": [Function],
        "listItem": [Function],
        "paragraph": [Function],
        "strong": [Function],
        "thematicBreak": [Function],
      }
    }
    skipHtml={false}
    source="
# Hello World

Goto [Underbelly](https://underbelly.is)
"
    transformLinkUri={[Function]}
  />
</glamorous(glamorous(div))>
`;
